File: app/src/main/java/com/niceplaces/niceplaces/adapters/PlacesAdapter.java
package com.niceplaces.niceplaces.adapters;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.niceplaces.niceplaces.R;
import com.niceplaces.niceplaces.activities.MapsActivity;
import com.niceplaces.niceplaces.models.Place;
import com.niceplaces.niceplaces.utils.ImageUtils;
import com.niceplaces.niceplaces.utils.MapUtils;

import java.util.List;

public class PlacesAdapter extends ArrayAdapter<Place> {

    private Context mContext;
    private GoogleMap mMap;

    public PlacesAdapter(Context context, int resource, List<Place> objects, GoogleMap map, LatLng position){
        super(context, resource, objects);
        mContext = context;
        mMap = map;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        final Place place = getItem(position);
        if(convertView == null){
            LayoutInflater layoutInflater = LayoutInflater.from(getContext());
            convertView = layoutInflater.inflate(R.layout.listview_places, parent, false);
        }
        ImageView imageViewPlaceImage = convertView.findViewById(R.id.imageview_place_image);
        ImageView imageViewPlaceStar = convertView.findViewById(R.id.imageview_place_star);
        ImageUtils.setAuthorIcon(place, imageViewPlaceStar);
        TextView textViewPlaceName = convertView.findViewById(R.id.textview_place_name);
        TextView textViewPlaceDistance = convertView.findViewById(R.id.textview_place_distance);
        ImageUtils.setImageViewWithGlide(mContext, place.mImage, imageViewPlaceImage);
        imageViewPlaceImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final LatLng center = new LatLng(place.mLatitude, place.mLongitude);
                GoogleMap.CancelableCallback callback = new GoogleMap.CancelableCallback() {
                    @Override
                    public void onFinish() {
                        mMap.animateCamera(CameraUpdateFactory.newCameraPosition(
                                new CameraPosition.Builder()
                                        .target(MapUtils.offset(mMap, center, 0, -150))
                                        .tilt(MapsActivity.DEFAULT_TILT)
                                        .zoom(mMap.getCameraPosition().zoom).build()));
                        place.mClusterItem.getMarker().showInfoWindow();
                        mMap.getUiSettings().setMapToolbarEnabled(true);
                    }

                    @Override
                    public void onCancel() {

                    }
                };
                float zoom = mMap.getCameraPosition().zoom;
                if (place.mClusterItem.isMarkerClustered()){
                    zoom = 18;
                }
                mMap.animateCamera(CameraUpdateFactory.newCameraPosition(
                        new CameraPosition.Builder()
                                .target(center)
                                .tilt(MapsActivity.DEFAULT_TILT)
                                .zoom(zoom).build()), callback);
            }
        });
        textViewPlaceName.setText(place.mName);
        textViewPlaceDistance.setText(Place.formatDistance(place.mDistance));
        /*Direction direction = new Direction(mPosition.latitude, mPosition.longitude, place.mLatitude, place.mLongitude);
        (new DirectionAsyncTask(mContext, textViewPlaceDistance)).execute(direction);*/
        return convertView;
    }
}

File: app/src/main/java/com/niceplaces/niceplaces/adapters/PlacesAdapter.kt
package com.niceplaces.niceplaces.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.CancelableCallback
import com.google.android.gms.maps.model.CameraPosition
import com.google.android.gms.maps.model.LatLng
import com.niceplaces.niceplaces.R
import com.niceplaces.niceplaces.activities.MapsActivity
import com.niceplaces.niceplaces.dao.DaoPlaces
import com.niceplaces.niceplaces.models.Place
import com.niceplaces.niceplaces.utils.ImageUtils
import com.niceplaces.niceplaces.utils.MapUtils
import com.niceplaces.niceplaces.utils.MyRunnable
import org.json.JSONObject

class PlacesAdapter(val context: Context, private val map: GoogleMap) :
    ListAdapter<Place, PlacesAdapter.PlacesViewHolder>(PlacesDiffCallback) {

    class PlacesViewHolder(itemView: View, val context: Context, val map: GoogleMap) :
        RecyclerView.ViewHolder(itemView) {
        private val textViewPlaceName = itemView.findViewById<TextView>(R.id.textview_place_name)
        private val textViewPlaceDistance =
            itemView.findViewById<TextView>(R.id.textview_place_distance)
        private val imageViewPlaceImage =
            itemView.findViewById<ImageView>(R.id.imageview_place_image)
        private val imageViewPlaceStar = itemView.findViewById<ImageView>(R.id.imageview_place_star)
        private var currentPlace: Place? = null

        fun bind(place: Place) {
            currentPlace = place
            textViewPlaceName.text = place.mName
            textViewPlaceDistance.text = Place.formatDistance(place.mDistance)
            ImageUtils.setAuthorIcon(place, imageViewPlaceStar)
            if (place.mImage != "") {
                ImageUtils.setImageViewWithGlide(context, place.mImage, imageViewPlaceImage)
            } else {
                if (place.mWikiUrl != "") {
                    val pageName = place.mWikiUrl?.let {
                        place.mWikiUrl?.substring(it.lastIndexOf('/') + 1)
                    }
                    if (pageName != null) {
                        DaoPlaces.getWikipediaData(context, pageName, false,
                            object : MyRunnable() {
                                override fun run() {
                                    val data = JSONObject(this.wikipediaData)
                                    if (place.mImage == "") {
                                        data.getJSONObject("thumbnail").getString("source").let {
                                            ImageUtils.setImageViewFromURL(
                                                context,
                                                it,
                                                imageViewPlaceImage
                                            )
                                        }
                                    }
                                }
                            },
                            Runnable() {

                            })
                    }
                }
            }
            itemView.setOnClickListener {
                val center = LatLng(place.mLatitude, place.mLongitude)
                val callback: CancelableCallback = object : CancelableCallback {
                    override fun onFinish() {
                        map.animateCamera(
                            CameraUpdateFactory.newCameraPosition(
                                CameraPosition.Builder()
                                    .target(MapUtils.offset(map, center, 0, -150))
                                    .tilt(MapsActivity.DEFAULT_TILT.toFloat())
                                    .zoom(map.cameraPosition.zoom).build()
                            )
                        )
                        place.mClusterItem?.marker?.showInfoWindow()
                        map.uiSettings.isMapToolbarEnabled = true
                    }

                    override fun onCancel() {}
                }
                var zoom = map.cameraPosition.zoom
                if (place.mClusterItem?.isMarkerClustered!!) {
                    zoom = 18f
                }
                map.animateCamera(
                    CameraUpdateFactory.newCameraPosition(
                        CameraPosition.Builder()
                            .target(center)
                            .tilt(MapsActivity.DEFAULT_TILT.toFloat())
                            .zoom(zoom).build()
                    ), callback
                )
            }
        }
    }

    /* Creates and inflates view and return PlaceViewHolder. */
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PlacesViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.listview_places, parent, false)
        return PlacesViewHolder(view, context, map)
    }

    /* Gets current place and uses it to bind view. */
    override fun onBindViewHolder(holder: PlacesViewHolder, position: Int) {
        val place = getItem(position)
        holder.bind(place)
    }

}

object PlacesDiffCallback : DiffUtil.ItemCallback<Place>() {
    override fun areItemsTheSame(oldItem: Place, newItem: Place): Boolean {
        return oldItem == newItem
    }

    override fun areContentsTheSame(oldItem: Place, newItem: Place): Boolean {
        return oldItem.mName == newItem.mName
    }
}